core/state_processor.go

```diff
@@ -126,18 +156,12 @@ func (p *StateProcessor) Process(block *types.Block, statedb *state.StateDB, cfg
 
 		// if less than maxCount slots were moved, move to the next account
 		for count < maxMovedCount {
-			statedb.Database().SetCurrentAccountHash(accIt.Hash())
-
 			acc, err := snapshot.FullAccount(accIt.Account())
 			if err != nil {
 				log.Error("Invalid account encountered during traversal", "error", err)
 				return nil, nil, 0, err
 			}
-			addr := rawdb.ReadPreimage(statedb.Database().DiskDB(), accIt.Hash())
-			if len(addr) == 0 {
-				panic(fmt.Sprintf("%x %x %v", addr, accIt.Hash(), acc))
-			}
-			vkt.SetStorageRootConversion(addr, common.BytesToHash(acc.Root))
+			vkt.SetStorageRootConversion(migrdb.GetCurrentAccountAddress().Bytes(), common.BytesToHash(acc.Root))
```

Comment:
Note that our invariant is that `GetCurrentAccountAddress()` always return the account that we're migration now.
This is always true, even between blocks (e.g: in a large contract where migrating the state might require multiple blocks).

The good part here is that we don't have to worry about "jumping back" into the file to read again the account preimage, if we're in the middle of migrating it's storage slots.